src/hotspot/share/opto/divnode.cpp

```diff
@@ -447,6 +447,81 @@ static Node *transform_long_divide( PhaseGVN *phase, Node *dividend, jlong divis
   return q;
 }
 
+template <typename TypeClass>
+Node* make_and(Node* a, Node* b);
+
+template <>
+Node* make_and<TypeLong>(Node* a, Node* b) {
+  return new AndLNode(a, b);
+}
+
+template <>
+Node* make_and<TypeInt>(Node* a, Node* b) {
+  return new AndINode(a, b);
+}
+
+template <typename TypeClass>
+Node* make_urshift(Node* a, Node* b);
+
+template <>
+Node* make_urshift<TypeLong>(Node* a, Node* b) {
+  return new URShiftLNode(a, b);
+}
+
+template <>
+Node* make_urshift<TypeInt>(Node* a, Node* b) {
+  return new URShiftINode(a, b);
+}
+
+template <typename TypeClass, typename Signed>
+Node* unsigned_div_ideal(PhaseGVN* phase, bool can_reshape, Node* div) {
+  // Check for dead control input
+  if (div->in(0) && div->remove_dead_region(phase, can_reshape)) {
+    return div;
+  }
+  // Don't bother trying to transform a dead node
+  if (div->in(0) && div->in(0)->is_top()) {
+    return nullptr;
+  }
+
+  const Type* t = phase->type(div->in(2));
+  if (t == TypeClass::ONE) { // Identity?
+    return nullptr;          // Skip it
+  }
+
+  const TypeClass* tl = t->is<TypeClass>();
+  if (!tl) {
+    return nullptr;
+  }
+
+  // Check for useless control input
+  // Check for excluding div-zero case
+  if (div->in(0) && (tl->_hi < 0 || tl->_lo > 0)) {
+    div->set_req(0, nullptr); // Yank control input
+    return div;
+  }
+
+  if (!tl->is_con()) {
+    return nullptr;
+  }
+  Signed l = tl->get_con(); // Get divisor
+
+  if (l == 0) {
+    return nullptr; // Dividing by zero constant does not idealize
+  }
+
+  if (l == min_jint) {
+    return nullptr;
+  }
+
+  if (is_power_of_2(l)) {
+    return make_urshift<TypeClass>(div->in(1), phase->intcon(log2i_graceful(l)));
```

Comment:
Are we testing optimizations like these with random constants somewhere, and comparing it to the interpreter results?

Reply:
https://github.com/openjdk/jdk/pull/22061/files#diff-48b0b8da547a3fe6aae9ea3ef20b4d708e47f2332ff6884478336f39d9eb9459R82 and https://github.com/openjdk/jdk/pull/22061/files#diff-24679e6505fe23e8a3ba73decaaf97896899c0a10956c437b8721fca33706ee2R82 should cover this I think. The containing method is marked with @DontCompile.

Reply:
You could use a similar trick with the constant method handles, as here:
https://github.com/openjdk/jdk/pull/21521/files#diff-d69ed849846cce04a18fe13fb35cd975ad533f0ef76d923745d97bdb27db7073

Reply:
Tried to address this in my latest push.